modules/contact/src/constraints/ComputeWeightedGapLMMechanicalContact.C

```diff
@@ -51,41 +54,107 @@ ComputeWeightedGapLMMechanicalContact::ComputeWeightedGapLMMechanicalContact(
+}
+
+void
+ComputeWeightedGapLMMechanicalContact::computeQpProperties()
+{
+  if (_has_primary)
   {
-    case Moose::MortarType::Lower:
+    ADRealVectorValue gap_vec = _phys_points_primary[_qp] - _phys_points_secondary[_qp];
+
+    if (_displaced)
     {
-      if (_has_primary)
-      {
-        DualRealVectorValue gap_vec = _phys_points_primary[_qp] - _phys_points_secondary[_qp];
-
-        if (_displaced)
-        {
-          gap_vec(0).derivatives() =
-              _primary_disp_x[_qp].derivatives() - _secondary_disp_x[_qp].derivatives();
-          gap_vec(1).derivatives() =
-              _primary_disp_y[_qp].derivatives() - _secondary_disp_y[_qp].derivatives();
-        }
-
-        mooseAssert(_normals.size() == (_interpolate_normals ? _test[_i].size()
-                                                             : _lower_secondary_elem->n_nodes()),
-                    "Making sure that _normals is the expected size");
-        auto gap = gap_vec * _normals[_normal_index];
-
-        return _test[_i][_qp] * gap;
-      }
-      else
-        return std::numeric_limits<Real>::quiet_NaN();
+      gap_vec(0).derivatives() =
+          _primary_disp_x[_qp].derivatives() - _secondary_disp_x[_qp].derivatives();
+      gap_vec(1).derivatives() =
+          _primary_disp_y[_qp].derivatives() - _secondary_disp_y[_qp].derivatives();
     }
 
-    default:
-      return 0;
+    _qp_gap = gap_vec * (_normals[_normal_index] * _JxW_msm[_qp] * _coord[_qp]);
+  }
+  else
+    _qp_gap = std::numeric_limits<Real>::quiet_NaN();
+}
+
+void
+ComputeWeightedGapLMMechanicalContact::computeQpIProperties()
+{
+  mooseAssert(_normals.size() ==
+                  (_interpolate_normals ? _test[_i].size() : _lower_secondary_elem->n_nodes()),
+              "Making sure that _normals is the expected size");
+
+  const auto * const node = _lower_secondary_elem->node_ptr(_i);
+  _node_to_weighted_gap[node] += _test[_i][_qp] * _qp_gap;
+}
+
+void
+ComputeWeightedGapLMMechanicalContact::residualSetup()
+{
+  _node_to_weighted_gap.clear();
+}
+
+void
+ComputeWeightedGapLMMechanicalContact::jacobianSetup()
+{
+  residualSetup();
+}
+
+void
+ComputeWeightedGapLMMechanicalContact::computeResidual(const Moose::MortarType mortar_type)
+{
+  if (mortar_type != Moose::MortarType::Lower)
+    return;
+
+  mooseAssert(_var, "LM variable is null");
+
+  for (_qp = 0; _qp < _qrule_msm->n_points(); _qp++)
+  {
+    computeQpProperties();
+    for (_i = 0; _i < _test.size(); ++_i)
+      computeQpIProperties();
+  }
+}
+
+void
+ComputeWeightedGapLMMechanicalContact::computeJacobian(const Moose::MortarType mortar_type)
+{
+  computeResidual(mortar_type);
+}
+
+void
+ComputeWeightedGapLMMechanicalContact::post()
+{
+  for (const auto & pr : _node_to_weighted_gap)
+  {
+    _weighted_gap_ptr = &pr.second;
+    onNode(pr.first);
   }
 }
+
+void
+ComputeWeightedGapLMMechanicalContact::onNode(const Node * const node)
```

Comment:
We could although I thought that in the future someone deriving from this class could make this method virtual and then do some number of unforeseen things in `onNode`. Although since this class is pretty specific already it is probably reasonable to expect that any derivers (you and frictional contact being the pair I have in my mind :smile: ) will be using this method to enforce constraints. So yea I'm on board for the name change